nestkernel/sonata_connector.cpp

```diff
@@ -0,0 +1,667 @@
+
+    if ( is_edge_id_name )
+    {
+      edge_id_grp_names.push_back( name );
+      ++num_edge_id_groups;
+    }
+  }
+
+  return num_edge_id_groups;
+}
+
+
+template < typename T >
+void
+SonataConnector::read_subset_( const H5::DataSet& dataset,
+  std::vector< T >& data_buf,
+  H5::PredType datatype,
+  hsize_t chunk_size,
+  hsize_t offset )
+{
+  try
+  {
+    H5::DataSpace mspace( 1, &chunk_size, NULL );
+    H5::DataSpace dspace = dataset.getSpace();
+    // Select hyperslab. H5S_SELECT_SET replaces any existing selection with this call
+    dspace.selectHyperslab( H5S_SELECT_SET, &chunk_size, &offset );
+    dataset.read( data_buf.data(), datatype, mspace, dspace );
+    mspace.close();
+    dspace.close();
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Unable to read datasets in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+}
+
+void
+SonataConnector::create_type_id_2_syn_spec_( DictionaryDatum edge_params )
+{
+  for ( auto it = edge_params->begin(); it != edge_params->end(); ++it )
+  {
+    const auto type_id = std::stoi( it->first.toString() );
+    auto d = getValue< DictionaryDatum >( it->second );
+    const auto syn_name = getValue< std::string >( ( *d )[ "synapse_model" ] );
+
+    // The following call will throw "UnknownSynapseType" if syn_name is not naming a known model
+    const index synapse_model_id = kernel().model_manager.get_synapse_model_id( syn_name );
+
+    set_synapse_params_( d, synapse_model_id, type_id );
+    type_id_2_syn_model_[ type_id ] = synapse_model_id;
+  }
+}
+
+void
+SonataConnector::set_synapse_params_( DictionaryDatum syn_dict, index synapse_model_id, int type_id )
+{
+  DictionaryDatum syn_defaults = kernel().model_manager.get_connector_defaults( synapse_model_id );
+  std::set< Name > skip_syn_params_ = {
+    names::weight, names::delay, names::min_delay, names::max_delay, names::num_connections, names::synapse_model
+  };
+
+  std::map< Name, std::shared_ptr< ConnParameter > > synapse_params; // TODO: Use unique_ptr/shared_ptr
+
+  for ( Dictionary::const_iterator default_it = syn_defaults->begin(); default_it != syn_defaults->end(); ++default_it )
+  {
+    const Name param_name = default_it->first;
+    if ( skip_syn_params_.find( param_name ) != skip_syn_params_.end() )
+    {
+      continue; // weight, delay or other not-settable parameter
+    }
+
+    if ( syn_dict->known( param_name ) )
+    {
+      synapse_params[ param_name ] = std::shared_ptr< ConnParameter >(
+        ConnParameter::create( ( *syn_dict )[ param_name ], kernel().vp_manager.get_num_threads() ) );
+    }
+  }
+
+  // Now create dictionary with dummy values that we will use to pass settings to the synapses created. We
+  // create it here once to avoid re-creating the object over and over again.
+  for ( thread tid = 0; tid < kernel().vp_manager.get_num_threads(); ++tid )
+  {
+    type_id_2_syn_spec_[ type_id ].push_back( synapse_params ); // DO WE NEED TO DEFINE THIS PER THREAD???
```

Comment:
This comment looks like one that needs to be resolved before merge.

Reply:
Thanks for pointing this out, the comment is now resolved. Please take a look at the new implementation.